File: gate_web/python/order.py
from __future__ import annotations

import time
from enum import Enum
from dataclasses import dataclass
from typing import Optional


class Side(str, Enum):
    LONG = "long"
    SHORT = "short"


class OrderType(str, Enum):
    MARKET = "market"
    LIMIT = "limit"


class TimeInForce(str, Enum):
    GTC = "gtc"  # good till cancel
    IOC = "ioc"  # immediate or cancel
    POC = "poc"  # pending or cancel (post-only)


@dataclass
class OrderParams:
    """Gate.io futures order parameters.

    Gate uses signed size: positive = long, negative = short.
    The ``side`` field controls the sign automatically.
    """

    side: Side
    size: int  # number of contracts (unsigned, sign derived from side)
    order_type: OrderType = OrderType.MARKET
    price: Optional[float] = None  # required for limit; "0" sent for market
    tif: TimeInForce = TimeInForce.GTC
    reduce_only: bool = False
    text: Optional[str] = None  # custom label, e.g. "t-1234567890"

    def __post_init__(self):
        if self.order_type == OrderType.LIMIT and self.price is None:
            raise ValueError("price is required for limit orders")

    def to_body(self, contract: str) -> dict:
        """Build the JSON body for the Gate.io futures order API.

        contract: e.g. "BTC_USDT"
        """
        signed_size = self.size if self.side == Side.LONG else -self.size
        price_str = str(self.price) if self.price is not None else "0"
        text = self.text or f"t-{int(time.time() * 1000)}"
        return {
            "contract": contract,
            "size": str(signed_size),
            "price": price_str,
            "order_type": self.order_type.value,
            "tif": self.tif.value,
            "text": text,
            "reduce_only": self.reduce_only,
        }

    @staticmethod
    def close_body(contract: str, size: int, price: Optional[float] = None) -> dict:
        """Build a reduce-only order to close ``size`` contracts.

        size: positive = close short (buy), negative = close long (sell).
        price: limit price, or None/0 for market.
        """
        return {
            "contract": contract,
            "size": str(size),
            "price": str(price) if price else "0",
            "order_type": "market",
            "tif": "ioc",
            "text": f"t-{int(time.time() * 1000)}",
            "reduce_only": True,
        }

File: gate_web/rust/src/order.rs
use serde::{Deserialize, Serialize};
use std::time::{SystemTime, UNIX_EPOCH};

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum Side {
    Long,
    Short,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum OrderType {
    Market,
    Limit,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum TimeInForce {
    /// Good till cancel
    Gtc,
    /// Immediate or cancel
    Ioc,
    /// Pending or cancel (post-only)
    Poc,
}

#[derive(Debug, Clone)]
pub struct OrderParams {
    pub side: Side,
    /// Number of contracts (unsigned; sign is derived from `side`).
    pub size: i64,
    pub order_type: OrderType,
    /// Required for limit orders; ignored for market.
    pub price: Option<f64>,
    pub tif: TimeInForce,
    pub reduce_only: bool,
    /// Optional custom label, e.g. "t-1234567890".
    pub text: Option<String>,
}

impl OrderParams {
    pub fn to_body(&self, contract: &str) -> serde_json::Value {
        let signed_size = match self.side {
            Side::Long => self.size,
            Side::Short => -self.size,
        };
        let price_str = self
            .price
            .map(|p| format_price(p))
            .unwrap_or_else(|| "0".to_string());
        let text = self
            .text
            .clone()
            .unwrap_or_else(|| format!("t-{}", now_ms()));

        serde_json::json!({
            "contract": contract,
            "size": signed_size.to_string(),
            "price": price_str,
            "order_type": tif_or_type_to_str(self.order_type),
            "tif": tif_to_str(self.tif),
            "text": text,
            "reduce_only": self.reduce_only,
        })
    }

    /// Build a reduce-only market IOC body to close `signed_size` contracts.
    /// Positive = close short (buy), negative = close long (sell).
    pub fn close_body(contract: &str, signed_size: i64, price: Option<f64>) -> serde_json::Value {
        let price_str = price
            .filter(|p| *p > 0.0)
            .map(format_price)
            .unwrap_or_else(|| "0".to_string());
        serde_json::json!({
            "contract": contract,
            "size": signed_size.to_string(),
            "price": price_str,
            "order_type": "market",
            "tif": "ioc",
            "text": format!("t-{}", now_ms()),
            "reduce_only": true,
        })
    }
}

fn tif_or_type_to_str(t: OrderType) -> &'static str {
    match t {
        OrderType::Market => "market",
        OrderType::Limit => "limit",
    }
}

fn tif_to_str(t: TimeInForce) -> &'static str {
    match t {
        TimeInForce::Gtc => "gtc",
        TimeInForce::Ioc => "ioc",
        TimeInForce::Poc => "poc",
    }
}

fn now_ms() -> u128 {
    SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .map(|d| d.as_millis())
        .unwrap_or(0)
}

/// Format a price without trailing zeros or scientific notation.
/// Gate accepts decimal strings; matches Python's `str(float)` shape.
fn format_price(p: f64) -> String {
    let s = format!("{}", p);
    s
}

#[derive(Debug, Clone)]
pub struct OrderResponse {
    pub ok: bool,
    pub status: u16,
    pub data: serde_json::Value,
    pub elapsed_ms: f64,
}

impl OrderResponse {
    /// Inner data dict from `data.data` if present, else empty map.
    pub fn inner(&self) -> &serde_json::Value {
        self.data.get("data").unwrap_or(&serde_json::Value::Null)
    }

    pub fn message(&self) -> &str {
        self.data
            .get("message")
            .and_then(|v| v.as_str())
            .unwrap_or("")
    }

    pub fn label(&self) -> &str {
        self.data
            .get("label")
            .and_then(|v| v.as_str())
            .unwrap_or("")
    }
}
